app/questionnaire/completeness.py

```diff
@@ -0,0 +1,161 @@
+        return location in self.completed_blocks
+
+    def all_sections_complete(self):
+        return all(state in [self.COMPLETED, self.SKIPPED] for state in self._get_all_section_states())
+
+    def any_section_complete(self):
+        return any(state == self.COMPLETED for state in self._get_all_section_states())
+
+    def get_state_for_section(self, section):
+        if isinstance(section, str):
+            # lookup section by section ID
+            section = self.schema.get_section(section)
+
+        group_states = [
+            self.get_state_for_group(group) for group in section['groups']
+        ]
+
+        if not group_states:
+            return self.NOT_STARTED
+
+        if all(state == self.SKIPPED for state in group_states):
+            return self.SKIPPED
+
+        if all(state in [self.COMPLETED, self.SKIPPED] for state in group_states):
+            return self.COMPLETED
+
+        if any(state in [self.STARTED, self.COMPLETED] for state in group_states):
+            return self.STARTED
+
+        return self.NOT_STARTED
+
+    def get_state_for_group(self, group, group_instance=None):
+        if isinstance(group, str):
+            # lookup group by group ID
+            group = self.schema.get_group(group)
+
+        if self._should_skip_group(group):
+            return self.SKIPPED
+
+        block_states = [
+            state for location, state in
+            self._get_block_states_for_group(group, group_instance)
+        ]
+
+        if self.routing_path and not block_states:
+            # the questionnaire has been started but none of the blocks for this group
+            # are on the current routing path
+            return self.SKIPPED
+
+        elif any(block_states):
+            if all(block_states):
+                return self.COMPLETED
+
+            return self.STARTED
+
+        return self.NOT_STARTED
+
+    def get_first_incomplete_location_in_survey(self):
+        incomplete_locations = (
+            self.get_first_incomplete_location_in_section(section)
+            for section in self._get_all_question_sections()
+        )
+        return next(filter(None, incomplete_locations), None)
+
+    def get_first_incomplete_location_in_section(self, section):
+        incomplete_locations = (
+            self.get_first_incomplete_location_in_group(group)
+            for group in section['groups']
+        )
+        return next(filter(None, incomplete_locations), None)
+
+    def get_first_incomplete_location_in_group(self, group, group_instance=None):
+        if self._should_skip_group(group):
+            return
+
+        incomplete_locations = (
+            location for location, is_complete in
+            self._get_block_states_for_group(
+                group, group_instance=group_instance)
+            if not is_complete
+        )
+        return next(incomplete_locations, None)
+
+    def is_block_valid_on_path(self, location, block):
+        if block['type'] == 'Question':
```

Comment:
if type isn't `Question` this will return `None`?
Should it not return `false`?

Reply:
Done